### PR TITLE
Stripe Payments Scripts and Test

### PR DESCRIPTION
**Pull request by chinazor-allen**
#### What does this PR do?
Allows users to process stripe payments actions

#### Description of Task to be completed?
 -  As a developer, I want to be able to call actions such as balance, bank accounts, cards, etc using stripe payment service so that data gotten can be useful

#### What are the relevant JIRA stories?
`SL-32`

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+  "dependencies": {
+    "request": "^2.83.0",
+    "stripe": "^5.4.0",
+    "syncano-server": "beta"
+  },
+  "scripts": {
+    "test": "mocha --compilers js:babel-core/register test/*.js --timeout 15000"
+  },
+  "devDependencies": {
+    "babel-cli": "^6.26.0",
```
Comment (adamwardecki) [inline package.json]: Is this package necessary?

File: socket.yml [hunk 2]
```diff
@@ -0,0 +1,854 @@
+name: stripe-payments
+description: Stripe offers payment solutions for businesses.
```
Comment (adamwardecki) [inline socket.yml:2]: Bit more verbose description: link to stripe website, what endpoints the socket covers, link to stripe docs etc.

File: socket.yml [hunk 3]
```diff
@@ -0,0 +1,854 @@
...
+        example: |
+          {
+            "chargeParameter":
+              {"currency": "usd",
+                "source": "tok_mastercard",
+                "amount": 500,
+                "capture":"false"
+              }
+          }
+      chargeID:
```
Comment (adamwardecki) [inline socket.yml:189]: Add info that this param is required for retrieve, update and shouldn't be passed for create

File: socket.yml [hunk 4]
```diff
@@ -0,0 +1,854 @@
...
+          description: Failed
+          example: |
+            {
+              "type": "StripeInvalidRequestError",
+              "message": "Received unknown parameter: id",
+              "statusCode": 400
+            }
+  
+  customers/customer:
+    description: Create, Retrieve, Update and Delete  actions for customer.
```
Comment (adamwardecki) [inline socket.yml]: Double space between "Delete" and "actions"

File: socket.yml [hunk 5]
```diff
@@ -0,0 +1,854 @@
...
+
+  tokens/token:
+    description: Creates a single use token that wraps the details of a credit card.
+    parameters:
+      tokenParams:
+        type: object
+        description: Parameters for creating token 
+        long_description: | 
+          Parameters provided must be either a card, customer, pii data, or bank account to create a token.
+          Reference Stripe API reference documentation.
```
Comment (adamwardecki) [inline socket.yml:608]: "Reference Stripe API reference documentation." this could be phrased better

File: socket.yml [hunk 6]
```diff
@@ -0,0 +1,854 @@
...
+          description: Failed
+          example: |
+            {
+              "type": "StripeInvalidRequestError",
+              "message": "You must supply either a card, customer, pii data, or bank account to create a token.",
+              "statusCode": 400
+            }
+
+  bankAccounts/bankAccount:
+    description: Create, Retrieve, Update and Delete  actions for bank account.
```
Comment (adamwardecki) [inline socket.yml]: double space

File: socket.yml [hunk 7]
```diff
@@ -0,0 +1,854 @@
...
+          description: Failed
+          example: |
+            {
+              "type": "StripeInvalidRequestError",
+              "message": "Missing required param: source.",
+              "statusCode": 400
+            }
+
+  bankAccounts/listAllBankAccts:
+    description: Returns  a list of the bank accounts belonging to a Customer.
```
Comment (adamwardecki) [inline socket.yml]: double space

File: socket.yml [hunk 8]
```diff
@@ -0,0 +1,854 @@
...
+          description: Failed
+          example: |
+            {
+              "type": "StripeInvalidRequestError",
+              "message": "The payment source card_1BZLlzJbpxha41ttuQ1XGBpE does not require validation.",
+              "statusCode": 400
+            }
+
+  cards/card:
+    description: Create, Retrieve, Update and Delete  actions for cards.
```
Comment (adamwardecki) [inline socket.yml]: double space

File: src/balance/listAllBalance.js [hunk 9]
```diff
@@ -0,0 +1,30 @@
+import Syncano from 'syncano-server';
+import stripePackage from 'stripe';
+
+export default async (ctx) => {
+  const { response, logger } = Syncano(ctx);
+  const log = logger('Socket scope');
```
Comment (adamwardecki) [inline src/balance/listAllBalance.js]: Remove the log if not used

File: src/balance/listAllBalance.js [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+  const stripe = stripePackage(ctx.config.STRIPE_SECRET_KEY);
+
+  try {
+    if (ctx.meta.request.REQUEST_METHOD === 'POST') {
```
Comment (adamwardecki) [inline src/balance/listAllBalance.js]: Should be `GET` method

File: src/balance/retrieveBalance.js [hunk 11]
```diff
@@ -0,0 +1,27 @@
+import Syncano from 'syncano-server';
+import stripePackage from 'stripe';
+
+export default async (ctx) => {
+  const { response, logger } = Syncano(ctx);
+
+  const log = logger('Socket scope');
```
Comment (adamwardecki) [inline src/balance/retrieveBalance.js]: remove if not used

File: src/balance/retrieveBalanceTransaction.js [hunk 12]
```diff
@@ -0,0 +1,32 @@
+import Syncano from 'syncano-server';
+import stripePackage from 'stripe';
+
+export default async (ctx) => {
+  const { response, logger } = Syncano(ctx);
+  const log = logger('Socket scope');
```
Comment (adamwardecki) [inline src/balance/retrieveBalanceTransaction.js]: same as above and concerns all the script files

File: src/bankAccounts/bankAccount.js [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+    } else if (requestMethod === 'DELETE') {
+      const deleteBankAccount = await stripe.customers.deleteSource(customerID, bankAcctID);
+      return response.json({
+        message: 'Bank Account Deleted',
+        statusCode: 200,
+        data: deleteBankAccount,
+      });
+    }
+    throw new Error(
+      'Make sure to use `POST`, `GET`, `PUT` & `DELETE` request method for creating, ' +
```
Comment (adamwardecki) [inline src/bankAccounts/bankAccount.js]: use template literals for multiline strings

File: src/bankAccounts/bankAccount.js [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+      });
+    } else if (requestMethod === 'GET') {
+      const retrieveBankAcct = await stripe.customers.retrieveSource(customerID, bankAcctID);
+      return response.json({
+        message: 'Bank Account Retrieved',
+        statusCode: 200,
+        data: retrieveBankAcct,
+      });
+    } else if (requestMethod === 'PUT') {
+      const updateBankAccount = await stripe.customers.updateCard(
```
Comment (adamwardecki) [inline src/bankAccounts/bankAccount.js]: Looks like a bug in Stripe docs. It should be `updateSource`. Confirm and let me know

File: src/bankAccounts/listAllBankAccts.js [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+
+export default async (ctx) => {
+  const { response, logger } = Syncano(ctx);
+  const log = logger('Socket scope');
+  const stripe = stripePackage(ctx.config.STRIPE_SECRET_KEY);
+  const requestMethod = ctx.meta.request.REQUEST_METHOD;
+  const { customerID, bankAcctParams } = ctx.args;
+
+  try {
+    if (requestMethod === 'POST') {
```
Comment (adamwardecki) [inline src/bankAccounts/listAllBankAccts.js]: should be  `GET` request type for list methods

File: src/balance/retrieveBalance.js [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+  const stripe = stripePackage(ctx.config.STRIPE_SECRET_KEY);
+
+  try {
+    if (ctx.meta.request.REQUEST_METHOD === 'GET') {
```
Comment (adamwardecki) [inline src/balance/retrieveBalance.js]: The request type check is made in multiple scripts. Abstract it into a utility function (along with the throw error statement) and move before the try catch blocks in all the scrtips.

File: src/balance/retrieveBalanceTransaction.js [hunk 17]
```diff
@@ -0,0 +1,32 @@
...
+          data: retrieveBalanceTransaction,
+        },
+        200,
+      );
+    }
+    throw new Error(
+      'Make sure to use `GET` request method for retrieving balance transaction.',
+      400,
+    );
+  } catch (err) {
```
Comment (adamwardecki) [inline src/balance/retrieveBalanceTransaction.js]: catch ({ type, message, statusCode }) { response.json({ type, message, statusCode })}

File: src/cards/listAllCards.js [hunk 18]
```diff
@@ -0,0 +1,28 @@
...
+
+export default async (ctx) => {
+  const { response, logger } = Syncano(ctx);
+  const log = logger('Socket scope');
+  const stripe = stripePackage(ctx.config.STRIPE_SECRET_KEY);
+  const requestMethod = ctx.meta.request.REQUEST_METHOD;
+  const { customerID, cardParams } = ctx.args;
+
+  try {
+    if (requestMethod === 'POST') {
```
Comment (adamwardecki) [inline src/cards/listAllCards.js]: `GET` for listing

File: src/charge/listAllCharges.js [hunk 19]
```diff
@@ -0,0 +1,26 @@
+import Syncano from 'syncano-server';
+import stripePackage from 'stripe';
+
+export default async (ctx) => {
+  const { response, logger } = Syncano(ctx);
+  const log = logger('Socket scope');
+  const stripe = stripePackage(ctx.config.STRIPE_SECRET_KEY);
+
+  try {
+    if (ctx.meta.request.REQUEST_METHOD === 'POST') {
```
Comment (adamwardecki) [inline src/charge/listAllCharges.js]: `GET`

File: src/customers/customer.js [hunk 20]
```diff
@@ -0,0 +1,53 @@
...
+      });
+    } else if (requestMethod === 'PUT') {
+      const updateStripeCustomer = await stripe.customers.update(customerID, customerParameter);
+      return response.json({
+        message: 'Customer updated successfully',
+        statusCode: 200,
+        data: updateStripeCustomer,
+      });
+    } else if (requestMethod === 'DELETE') {
+      const deleteStripeCustomer = await stripe.customers.update(customerID);
```
Comment (adamwardecki) [inline src/customers/customer.js]: Nope. Should be `stripe.customers.del()`.

File: src/customers/listAllCustomers.js [hunk 21]
```diff
@@ -0,0 +1,28 @@
...
+import stripePackage from 'stripe';
+
+export default async (ctx) => {
+  const { response, logger } = Syncano(ctx);
+
+  const log = logger('Socket scope');
+  const stripe = stripePackage(ctx.config.STRIPE_SECRET_KEY);
+
+  try {
+    if (ctx.meta.request.REQUEST_METHOD === 'POST') {
```
Comment (adamwardecki) [inline src/customers/listAllCustomers.js]: `GET`

File: src/disputes/listAllDisputes.js [hunk 22]
```diff
@@ -0,0 +1,26 @@
+import Syncano from 'syncano-server';
+import stripePackage from 'stripe';
+
+export default async (ctx) => {
+  const { response, logger } = Syncano(ctx);
+  const log = logger('Socket scope');
+  const stripe = stripePackage(ctx.config.STRIPE_SECRET_KEY);
+
+  try {
+    if (ctx.meta.request.REQUEST_METHOD === 'POST') {
```
Comment (adamwardecki) [inline src/disputes/listAllDisputes.js]: `GET`

File: test/balanceTest.js [hunk 23]
```diff
@@ -0,0 +1,134 @@
...
+  describe('Retrieve Balance', () => {
+    meta = generateMeta('balance/retrieveBalance');
+    it('returns an error if a `POST` request method is passed to Url', (done) => {
+      meta.request.REQUEST_METHOD = 'POST';
+      run('balance/retrieveBalance', {
+        meta,
+        config
+      }).then((response) => {
+        expect(response.data.message).to.equal('Make sure to use `GET` request '
+        + 'method for retrieving balance.');
```
Comment (adamwardecki) [inline test/balanceTest.js]: template literals

File: test/chargeTest.js [hunk 24]
```diff
@@ -0,0 +1,136 @@
+import { assert, expect } from 'chai';
+import { run, generateMeta } from 'syncano-test';
+
+const config = {
+  STRIPE_SECRET_KEY: process.env.STRIPE_SECRET_KEY
+};
+let meta = generateMeta('charge/charges');
+let args = {};
+let chargeID, customerID;
+// const source = '';
```
Comment (adamwardecki) [inline test/chargeTest.js]: remove unnecessary comments

**Review by adamwardecki — CHANGES_REQUESTED** [review]
Ok, looks goods in general. Please fix the following things:
- Refactor the function checking the request method  into a utility function that will also handle the error throwing
- use template literals for multiline strings
- Use `GET` http method for list endpoints

File: package.json (hunk 1 above)
Comment (chinazor-allen) [reply package.json]: Fixed

File: socket.yml (hunk 4 above)
Comment (chinazor-allen) [reply socket.yml]: fixed

File: socket.yml (hunk 6 above)
Comment (chinazor-allen) [reply socket.yml]: fixed

File: socket.yml (hunk 7 above)
Comment (chinazor-allen) [reply socket.yml]: fixed

File: socket.yml (hunk 8 above)
Comment (chinazor-allen) [reply socket.yml]: fixed

File: src/balance/listAllBalance.js (hunk 9 above)
Comment (chinazor-allen) [reply src/balance/listAllBalance.js]: fixed

File: src/balance/retrieveBalance.js (hunk 11 above)
Comment (chinazor-allen) [reply src/balance/retrieveBalance.js]: fixed

File: src/balance/retrieveBalanceTransaction.js (hunk 12 above)
Comment (chinazor-allen) [reply src/balance/retrieveBalanceTransaction.js]: fixed

File: src/balance/retrieveBalanceTransaction.js (hunk 17 above)
Comment (chinazor-allen) [reply src/balance/retrieveBalanceTransaction.js]: fixed

File: src/bankAccounts/bankAccount.js (hunk 14 above)
Comment (chinazor-allen) [reply src/bankAccounts/bankAccount.js]: it actually shows `updateCard`

Comment (adamwardecki) [reply src/bankAccounts/bankAccount.js]: @Andela-callen You mean it shows `updateCard` in the stripe docs? I know, but it seems to me it's a bug in their documentation.

File: socket.yml (hunk 3 above)
Comment (chinazor-allen) [reply socket.yml:189]: fixed

File: socket.yml (hunk 2 above)
Comment (chinazor-allen) [reply socket.yml:2]: fixed

File: socket.yml (hunk 5 above)
Comment (chinazor-allen) [reply socket.yml:608]: fixed

File: src/bankAccounts/bankAccount.js (hunk 14 above)
Comment (chinazor-allen) [reply src/bankAccounts/bankAccount.js]: yes, I think it is a bug too, because `stripe.customers.updateSource` works perfectly.

**Comment by adamwardecki** [comment]
The point of creating the `checkRequestType` helper method was that you wouldn't have to use the `if` clause on the endpoints where there's only one request type used.

Also, it would be better to do `if (methodType === 'PUT')` instead of `if (methodType === requestedMethodTypes[x])` because the latter adds unnecessary abstraction. 

Let's leave this as it is though and get back to it when doing a readme for this socket

**Review by adamwardecki — APPROVED** [review]

**Pull request merged**